lib/mail/delivery_recorder.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module Mail
+  class DeliveryRecorder
+    def self.setup!(enabled:)
+      Mail::Message.include(MessageExtension)
+      ActionMailer::Base.register_observer(new(enabled: enabled))
+    end
+
+    def initialize(enabled:)
+      @enabled = enabled
+    end
+
+    def delivered_email(mail)
+      return unless enabled?
+
+      response = mail.delivery_method.response
+
+      ApplicationRecord.transaction do
+        email = Email.create!(
+          id: response.id,
+          from: response.content["from_email"],
+          to: mail.original_to,
+          template_id: response.template["id"],
+          template_version: response.template["version"],
+          uri: response.uri,
+          personalisation: mail.header["personalisation"].unparsed_value,
+        )
+
+        email.associate_with(*User.where(email: email.to).to_a)
+
+        mail.associations.each do |object, name|
+          email.associate_with object, as: name
+        end
+      end
+    end
+
+  private
+
+    def enabled?
+      !!@enabled
+    end
+
+    module MessageExtension
+      def associations
+        @associations ||= []
+      end
+
+      def associate_with(object, as: nil) # rubocop:disable Naming/MethodParameterName
+        associations << [object, as]
+        self
+      end
```

Comment:
Please name this method something that does not get easily confused with the method that creates stuff in our database. Maybe `with_associated_object`?

Reply:
But this exists solely to create stuff in our database

Reply:
It exists to pass the data to the `delivery_recorder`. It so happens that the recorder uses the method which is named the same to store stuff in the database at a later, asynchronous time.

Reply:
It is not asynchronous - `deliver_later` executes a mailer method and delivery in a single go. Somehow `mail(...).associate_with(object)` reads better to me than `mail(...).with_associated_object`. I am also not certain how this could be confused with `the method that creates stuff in our database` as we don't really have any method like that anywhere. Happy to get it under voting though :)

Reply:
```
  def associate_with(*objects, as: nil) # rubocop:disable Naming/MethodParameterName
    objects.each do |object|
      Association.create!(
        email: self,
        object: object,
        name: as,
      )
    end
  end
```
  vs 
```
      def associate_with(object, as: nil) # rubocop:disable Naming/MethodParameterName
        associations << [object, as]
        self
      end
```
The first one is in the email class and 'creates stuff in our database'. The second one is here. It took me longer than I wanted to understand which one is called where exactly :P

Reply:
AH, right! In this case I'd rename the method on the Email record to `create_association_with` to make the database operation more prominent.